File: Data Structures & Algorithms/is-palindrome/submission-1.py
class Solution:
    def checkalpha(self, c):
        return (ord('Z')>=ord(c) and ord(c)>=ord('A')) or (ord('z')>=ord(c) and ord(c)>=ord('a')) or(ord('9')>=ord(c) and ord(c)>=ord('0'))
        

    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        l = 0
        r = n-1

        while l<r:
            while l<r and not self.checkalpha(s[l]):
                l = l+1
            while l<r and not self.checkalpha(s[r]):
                r -= 1
            if(s[l].lower() != s[r].lower()):
                return False;
            l = l+1
            r = r-1
        return True;
        
        

        

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void checking(TreeNode* root,long long low, long long high, int& flag){
        if(root==nullptr || flag ==0) return;
        if(root->val<=low || root->val >= high){
            flag = 0;
            return;
        }
        checking(root->left, low, root->val, flag);
        checking(root->right, root->val, high, flag);
    }
    bool isValidBST(TreeNode* root) {
        // all values in the left subtree should be lower than the root, and all values in the right sub tree should be higher than the root
        // so a range is set for both of them
        
        if(root==nullptr) return true;
        int flag = 1;
        checking(root,LLONG_MIN, LLONG_MAX, flag);
        return flag;        
    }
};
